File: src/main/java/client/inventory/ModifyInventoryContext.java
package client.inventory;

import client.inventory.item.templates.ItemTemplate;
import client.inventory.operations.*;
import client.inventory.slots.ItemSlot;
import client.inventory.slots.ItemSlotBundle;
import constants.ItemConstants;
import lombok.Getter;
import managers.ItemManager;
import util.packet.PacketWriter;

import java.util.LinkedList;
import java.util.Map;
import java.util.Objects;
import java.util.Queue;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;

public class ModifyInventoryContext implements ModifyInventoryContextInterface {

    private final ItemInventoryType type;
    private final ItemInventory inventory;

    @Getter private final Queue<AbstractModifyInventoryOperation> operations;

    public ModifyInventoryContext(ItemInventoryType type, ItemInventory inventory) {
        this.type = type;
        this.inventory = inventory;

        operations = new LinkedList<>();
    }

    public ModifyInventoryContext(ItemInventory inventory) {
        this.type = ItemInventoryType.EQUIP;
        this.inventory = inventory;

        operations = new LinkedList<>();
    }

    @Override
    public void add(ItemSlot item) {
        if (item instanceof ItemSlotBundle) {
            ItemSlotBundle bundle = (ItemSlotBundle) item;
            if (bundle.getNumber() < 1) bundle.setNumber((short) 1);
            if (bundle.getMaxNumber() < 1) bundle.setMaxNumber((short) 1);

            while (bundle.getNumber() > bundle.getMaxNumber()) {
                bundle.setNumber((short) (bundle.getNumber() - bundle.getMaxNumber()));
                add(Objects.requireNonNull(ItemManager.getItem(item.getTemplateId())), bundle.getMaxNumber());
            }

            ItemSlotBundle mergeable = (ItemSlotBundle) inventory.getItems().values().stream().filter(i -> {
                ItemSlotBundle b = (ItemSlotBundle) i;
                return bundle.getNumber() + b.getNumber() <= b.getMaxNumber();
            }).filter(i -> i.equals(bundle)).findFirst().orElse(null);

            if (mergeable != null) {
                int quantity = bundle.getNumber() + mergeable.getNumber();
                short maxNumber = mergeable.getMaxNumber();

                if (quantity > maxNumber) {
                    int leftOver = quantity - maxNumber;
                    bundle.setNumber((short) leftOver);
                    mergeable.setNumber(maxNumber);
                    updateQuantity(mergeable);
                    add(bundle);
                    return;
                }

                mergeable.setNumber((short) (mergeable.getNumber() + bundle.getNumber()));
                updateQuantity(mergeable);
                return;
            }
        }
        short slot = (short) IntStream.range(1, inventory.getSlotMax())
                .filter(i -> {
                    short s = (short) i;
                    return inventory.getItems().get(s) == null;
                })
                .findFirst().orElse(0);

        set(slot, item);
    }

    @Override
    public void add(ItemTemplate item, short quantity) {
        ItemSlot i = item.toItemSlot();

        if (i instanceof ItemSlotBundle) {
            ((ItemSlotBundle) i).setNumber(quantity);
        }
        add(i);
    }

    @Override
    public void set(short slot, ItemSlot item) {
        inventory.getItems().put(slot, item);
        operations.add(new AddInventoryOperation(type, slot, item));
    }

    @Override
    public void set(short slot, ItemTemplate item, short quantity) {
        ItemSlot i = item.toItemSlot();

        if (i instanceof ItemSlotBundle) {
            ((ItemSlotBundle) i).setNumber(quantity);
        }
        set(slot, i);
    }

    @Override
    public void remove(short slot) {
        inventory.getItems().remove(slot);
        operations.add(new RemoveInventoryOperation(type, slot));
    }

    @Override
    public void remove(short slot, short quantity) {
        if (inventory.getItems().get(slot) instanceof ItemSlotBundle) {
            ItemSlotBundle bundle = (ItemSlotBundle) inventory.getItems().get(slot);
            if (!ItemConstants.isRechargeableItem(bundle.getTemplateId())) {
                if (quantity > 0) {
                    bundle.setNumber((short) (bundle.getNumber() - quantity));
                    bundle.setNumber((short) Math.max((short) 0, bundle.getNumber()));

                    if (bundle.getNumber() > 0) {
                        updateQuantity(bundle);
                        return;
                    }
                }
            }
        }
        remove(slot);
    }

    @Override
    public void remove(ItemSlot item) {
        remove(inventory.getItems().entrySet()
                .stream()
                .filter(entry -> entry.getValue() == item)
                .map(Map.Entry::getKey).findFirst().get());
    }

    @Override
    public void remove(ItemSlot item, short quantity) {
        remove(inventory.getItems().entrySet()
                .stream()
                .filter(entry -> entry.getValue() == item)
                .map(Map.Entry::getKey).findFirst().get(), quantity);
    }

    @Override
    public void remove(int id, short quantity) {
        final AtomicInteger removed = new AtomicInteger(0);
        inventory.getItems().values().stream()
                .filter(i -> i.getTemplateId() == id)
                .forEach(item -> {
                    System.out.println(item);
                    if (removed.get() >= quantity) return;
                    if (item instanceof ItemSlotBundle) {
                        ItemSlotBundle bundle = (ItemSlotBundle) item;
                        if (!ItemConstants.isRechargeableItem(bundle.getTemplateId())) {
                            int difference = quantity - removed.get();
                            if (bundle.getNumber() > difference) {
                                removed.addAndGet(difference);
                                bundle.setNumber((short) (bundle.getNumber() - difference));
                                updateQuantity(bundle);
                            } else {
                                removed.addAndGet(bundle.getNumber());
                                remove(bundle);
                            }
                        }
                    } else {
                        removed.incrementAndGet();
                        remove(item);
                    }
                });
    }

    @Override
    public void move(short from, short to) {
        ItemSlot item = inventory.getItems().get(from);

        if (item instanceof ItemSlotBundle) {
            ItemSlotBundle bundle = (ItemSlotBundle) item;
            if (!ItemConstants.isRechargeableItem(bundle.getTemplateId()) &&
                    inventory.getItems().containsKey(to) &&
                    inventory.getItems().get(to) instanceof ItemSlotBundle) {
                ItemSlotBundle existing = (ItemSlotBundle) inventory.getItems().get(to);
                if (bundle.equals(existing)) {
                    int quantity = bundle.getNumber() + existing.getNumber();
                    int max = existing.getMaxNumber();

                    if (quantity > max) {
                        int left = quantity - max;

                        bundle.setNumber((short) left);
                        existing.setNumber((short) max);
                        updateQuantity(bundle);
                    } else {
                        existing.setNumber((short) quantity);
                        remove(bundle);
                    }

                    updateQuantity(existing);
                    return;
                }
            }
        }
        inventory.getItems().remove(from);
        if (inventory.getItems().containsKey(to)) {
            inventory.getItems().put(from, inventory.getItems().get(to));
        }
        inventory.getItems().put(to, item);
        operations.add(new MoveInventoryOperation(type, from, to));
    }

    @Override
    public ItemSlotBundle take(short slot, short quantity) {
        ItemSlotBundle bundle = (ItemSlotBundle) inventory.getItems().get(slot);

        ItemSlotBundle newBundle = new ItemSlotBundle();
        newBundle.setTemplateId(bundle.getTemplateId());
        newBundle.setNumber(quantity);
        newBundle.setMaxNumber(bundle.getMaxNumber());
        newBundle.setAttribute(bundle.getAttribute());
        newBundle.setTitle(bundle.getTitle());
        newBundle.setExpire(bundle.getExpire());
        newBundle.setCashItemSN(bundle.getCashItemSN());

        remove(bundle, quantity);
        return newBundle;
    }

    @Override
    public ItemSlotBundle take(ItemSlotBundle bundle, short quantity) {
        return take(bundle.getTemplateId(), quantity);
    }

    @Override
    public ItemSlotBundle take(int id, short quantity) {
        return take(inventory.getItems().entrySet()
                .stream()
                .filter(entry -> entry.getValue().getTemplateId() == id)
                .map(Map.Entry::getKey).findFirst().get(), quantity);
    }

    @Override
    public void update(short slot) {
        ItemSlot item = inventory.getItems().get(slot);

        remove(slot);
        set(slot, item);
    }

    @Override
    public void update(ItemSlot slot) {
        update(inventory.getItems().entrySet()
                .stream()
                .filter(entry -> entry.getValue() == slot)
                .map(Map.Entry::getKey).findFirst().get());
    }

    @Override
    public void encode(PacketWriter pw) {
        pw.writeByte((byte) operations.size());
        operations.forEach(operation -> operation.encode(pw));
    }

    private void updateQuantity(ItemSlot item) {
        updateQuantity(inventory.getItems().entrySet()
                .stream()
                .filter(entry -> entry.getValue() == item)
                .map(Map.Entry::getKey).findFirst().get());
    }

    private void updateQuantity(short slot) {
        operations.add(new UpdateQuantityInventoryOperation(type, slot, ((ItemSlotBundle) inventory.getItems().get(slot)).getNumber()));
    }
}

File: src/main/java/client/inventory/ModifyInventoryContext.kt
package client.inventory

import client.inventory.item.templates.ItemTemplate
import client.inventory.operations.*
import client.inventory.item.slots.ItemSlot
import client.inventory.item.slots.ItemSlotBundle
import constants.ItemConstants.isRechargeableItem
import managers.ItemManager.getItem
import util.packet.PacketWriter
import java.util.*
import java.util.stream.IntStream
import kotlin.math.max

class ModifyInventoryContext : ModifyInventoryContextInterface {

    private val type: ItemInventoryType
    private val inventory: ItemInventory
    val operations: Queue<AbstractModifyInventoryOperation>

    constructor(type: ItemInventoryType, inventory: ItemInventory) {
        this.type = type
        this.inventory = inventory
        operations = LinkedList()
    }

    constructor(inventory: ItemInventory) {
        type = ItemInventoryType.EQUIP
        this.inventory = inventory
        operations = LinkedList()
    }

    override fun add(item: ItemSlot) {
        if (item is ItemSlotBundle) {
            if (item.number < 1) item.number = 1.toShort()
            if (item.maxNumber < 1) item.maxNumber = 1.toShort()

            val template = getItem(item.templateId) ?: return
            while (item.number > item.maxNumber) {
                item.number = (item.number - item.maxNumber).toShort()
                add(template, item.maxNumber)
            }

            val mergeable: ItemSlotBundle? = inventory.items.values.stream()
                    .filter { it is ItemSlotBundle }
                    .filter {
                        val b = it as ItemSlotBundle
                        item.number + b.number <= b.maxNumber
                    }.filter { it == item }
                    .findFirst().orElse(null) as ItemSlotBundle?

            if (mergeable != null) {
                val quantity = item.number + mergeable.number
                val maxNumber = mergeable.maxNumber

                if (quantity > maxNumber) {
                    val leftOver = quantity - maxNumber
                    item.number = leftOver.toShort()
                    mergeable.number = maxNumber
                    updateQuantity(mergeable)
                    add(item)
                    return
                }

                mergeable.number = (mergeable.number + item.number).toShort()
                updateQuantity(mergeable)
                return
            }
        }

        val slot = IntStream.range(1, inventory.slotMax.toInt())
                .filter {
                    val s = it.toShort()
                    inventory.items[s] == null
                }
                .findFirst().orElse(0).toShort()
        set(slot, item)
    }

    override fun add(item: ItemTemplate, quantity: Short) {
        val i = item.toItemSlot()
        if (i is ItemSlotBundle) {
            i.number = quantity
        }
        add(i)
    }

    override fun set(slot: Short, item: ItemSlot) {
        inventory.items[slot] = item
        operations.add(AddInventoryOperation(type, slot, item))
    }

    override fun set(slot: Short, item: ItemTemplate, quantity: Short) {
        val i = item.toItemSlot()
        if (i is ItemSlotBundle) {
            i.number = quantity
        }
        set(slot, i)
    }

    override fun remove(slot: Short) {
        inventory.items.remove(slot)
        operations.add(RemoveInventoryOperation(type, slot))
    }

    override fun remove(slot: Short, quantity: Short) {
        val item = inventory.items[slot]

        if (item is ItemSlotBundle) {
            if (!isRechargeableItem(item.templateId)) {
                if (quantity > 0) {
                    item.number = (item.number - quantity).toShort()
                    item.number = max(0, item.number.toInt()).toShort()

                    if (item.number > 0) {
                        updateQuantity(item)
                        return
                    }
                }
            }
        }

        remove(slot)
    }

    override fun remove(item: ItemSlot) {
        remove(inventory.items.entries
                .stream()
                .filter { it.value === item }
                .map { it.key }.findFirst().get())
    }

    override fun remove(item: ItemSlot, quantity: Short) {
        remove(inventory.items.entries
                .stream()
                .filter { it.value === item }
                .map { it.key }.findFirst().get(), quantity)
    }

    override fun remove(id: Int, quantity: Short) {
        var removed = 0

        inventory.items.values.stream()
                .filter { it.templateId == id }
                .forEach {
                    println(it)
                    if (removed >= quantity) return@forEach

                    if (it is ItemSlotBundle) {
                        if (!isRechargeableItem(it.templateId)) {
                            val difference = quantity - removed

                            if (it.number > difference) {
                                removed += difference
                                it.number = (it.number - difference).toShort()
                                updateQuantity(it)
                            } else {
                                removed += it.number.toInt()
                                remove(it)
                            }
                        }
                    } else {
                        removed++
                        remove(it)
                    }
                }
    }

    override fun move(from: Short, to: Short) {
        val item = inventory.items[from] ?: return

        if (item is ItemSlotBundle) {
            if (!isRechargeableItem(item.templateId) &&
                    inventory.items.containsKey(to) &&
                    inventory.items[to] is ItemSlotBundle) {
                val existing = inventory.items[to] as ItemSlotBundle?

                if (item == existing) {
                    val quantity = item.number + existing.number
                    val max = existing.maxNumber.toInt()

                    if (quantity > max) {
                        val left = quantity - max
                        item.number = left.toShort()
                        existing.number = max.toShort()
                        updateQuantity(item)
                    } else {
                        existing.number = quantity.toShort()
                        remove(item)
                    }
                    updateQuantity(existing)
                    return
                }
            }
        }

        inventory.items.remove(from)
        inventory.items[to]?.let {
            inventory.items[from] = it
        }

        inventory.items[to] = item
        operations.add(MoveInventoryOperation(type, from, to))
    }

    override fun take(slot: Short, quantity: Short): ItemSlotBundle {
        val bundle = inventory.items[slot] as ItemSlotBundle
        val newBundle = ItemSlotBundle()

        newBundle.templateId = bundle.templateId
        newBundle.number = quantity
        newBundle.maxNumber = bundle.maxNumber
        newBundle.attribute = bundle.attribute
        newBundle.title = bundle.title
        newBundle.expire = bundle.expire
        newBundle.cashItemSN = bundle.cashItemSN

        remove(bundle, quantity)

        return newBundle
    }

    override fun take(bundle: ItemSlotBundle, quantity: Short): ItemSlotBundle {
        return take(bundle.templateId, quantity)
    }

    override fun take(id: Int, quantity: Short): ItemSlotBundle {
        return take(inventory.items.entries
                .stream()
                .filter { it.value.templateId == id }
                .map { it.key }.findFirst().get(), quantity)
    }

    override fun update(slot: Short) {
        val item = inventory.items[slot]
        remove(slot)
        set(slot, item!!)
    }

    override fun update(item: ItemSlot) {
        update(inventory.items.entries
                .stream()
                .filter { it.value === item }
                .map { it.key }.findFirst().get())
    }

    override fun encode(pw: PacketWriter) {
        pw.writeByte(operations.size.toByte())
        operations.forEach { it.encode(pw) }
    }

    private fun updateQuantity(item: ItemSlot) {
        updateQuantity(inventory.items.entries
                .stream()
                .filter { it.value === item }
                .map { it.key }.findFirst().get())
    }

    private fun updateQuantity(slot: Short) {
        operations.add(UpdateQuantityInventoryOperation(type, slot, (inventory.items[slot] as ItemSlotBundle?)!!.number))
    }
}